### PR TITLE
[Go] Concepts needed to solve the v2 two fer exercise

### PR DESCRIPTION
**Pull request by hans-d**
solves #230

### CODE REVIEW

File: languages/go/reference/exercise-concepts/two-fer.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+It's the first time most students work with go, and many have not yet been using proper documentation of the package and
+the exported function. Most of the time they leave the stub comments in the file.
+
+### Functions
+
+Everything is done via functions. In this case, it's also:
+
+- there is no concept of default values for paramaters, it's done in the body of the function
+- parameters are passed by-value, so in the scope of the function they can be altered
```
Comment (tehsphinx) [inline languages/go/reference/exercise-concepts/two-fer.md]: Very valuable!

File: languages/go/reference/exercise-concepts/two-fer.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+### Linting
+
+It's the first time most students work with go, and many have not yet used `golint`. Linting already provides some hints
+on what (not) to do, eg use of variable names, not using else if the the regular condition ends with return, and missing
+documentation on exported function (or in the wrong format). 
+
+### Comments
+
+It's the first time most students work with go, and many have not yet been using proper documentation of the package and
+the exported function. Most of the time they leave the stub comments in the file.
```
Comment (tehsphinx) [inline languages/go/reference/exercise-concepts/two-fer.md]: wondering if that is to be learned before anything else. Probably with other concepts like `strings` and `numbers` together.

**Comment by tehsphinx** [comment]
I'd leave this open for a few days so others can also review it

File: languages/go/reference/exercise-concepts/two-fer.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+## Concepts encountered while solving it
+
+### Code formatting
+
+It's the first time most students work with go, and many have not yet used `gofmt`.
+
+### Linting
+
+It's the first time most students work with go, and many have not yet used `golint`. Linting already provides some hints
+on what (not) to do, eg use of variable names, not using else if the the regular condition ends with return, and missing
```
Comment (kotp) [inline languages/go/reference/exercise-concepts/two-fer.md]:
```suggestion
on what (not) to do, e.g. use of variable names, not using else if the the regular condition ends with return, and missing
```

File: languages/go/reference/exercise-concepts/two-fer.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+documentation on exported function (or in the wrong format). 
+
+### Comments
+
+It's the first time most students work with go, and many have not yet been using proper documentation of the package and
+the exported function. Most of the time they leave the stub comments in the file.
+
+### Functions
+
+Everything is done via functions. In this case, it's also:
```
Comment (kotp) [inline languages/go/reference/exercise-concepts/two-fer.md]:
```suggestion
Everything is done via functions. In this case:
```

File: languages/go/reference/exercise-concepts/two-fer.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+The parameter passed is a string. For the default value, a check needs to be made if the string is empty.
+
+- how to do a idiomatic check on empty string
+- how to concatenate strings, either via simple `+` or via the fmt package
+
+
+### Variables / Assignment / Scope
+
+The parameters are passed by-value, so in the scope of the function they can be altered.
```
Comment (kotp) [inline languages/go/reference/exercise-concepts/two-fer.md]:
```suggestion
The parameters are passed by value, so in the scope of the function they can be altered.
```

File: languages/go/reference/exercise-concepts/two-fer.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+- scoping
+
+### Comparison / Conditionals
+
+The parameters needs to be checked if it's empty.
```
Comment (kotp) [inline languages/go/reference/exercise-concepts/two-fer.md]:
```suggestion
The parameters need to be checked if it's empty.
```

File: languages/go/reference/exercise-concepts/two-fer.md [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+## Exercise
+
+Given a name, return a string with the message: "One for X, one for me.", where the X can be provided and defaults to "you" if not given.
+
+## Concepts encountered while solving it
+
+### Code formatting
+
+It's the first time most students work with go, and many have not yet used `gofmt`.
```
Comment (kotp) [inline languages/go/reference/exercise-concepts/two-fer.md]:
```suggestion
It's the first time most students work with Go, and many have not yet used `gofmt`.
```

I believe the language is capitalized when referring to it.

File: languages/go/reference/exercise-concepts/two-fer.md [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+### Linting
+
+It's the first time most students work with go, and many have not yet used `golint`. Linting already provides some hints
```
Comment (kotp) [inline languages/go/reference/exercise-concepts/two-fer.md]:
```suggestion
It's the first time most students work with Go, and many have not yet used `golint`. Linting already provides some hints
```

File: languages/go/reference/exercise-concepts/two-fer.md [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+on what (not) to do, eg use of variable names, not using else if the the regular condition ends with return, and missing
+documentation on exported function (or in the wrong format). 
+
+### Comments
+
+It's the first time most students work with go, and many have not yet been using proper documentation of the package and
```
Comment (kotp) [inline languages/go/reference/exercise-concepts/two-fer.md]:
```suggestion
It's the first time most students work with Go, and many have not yet been using proper documentation of the package and
```

File: languages/go/reference/exercise-concepts/two-fer.md (hunk 1 above)
Comment (kotp) [reply languages/go/reference/exercise-concepts/two-fer.md]: Another reference for "by-value" that probably shouldn't have a hyphen.

File: languages/go/reference/exercise-concepts/two-fer.md (hunk 6 above)
Comment (tehsphinx) [reply languages/go/reference/exercise-concepts/two-fer.md]: Second suggestion is not really valid / misleading for Go as there are no optional parameters.

Actually in the case of `twofer` I think it should be:
```suggestion
The parameter needs to be checked if it's empty.
```

File: languages/go/reference/exercise-concepts/two-fer.md (hunk 3 above)
Comment (bitfield) [inline languages/go/reference/exercise-concepts/two-fer.md]: This isn't quite clear to me (what does 'use of variable names' mean? You should use them? You _shouldn't_ use them?)

File: languages/go/reference/exercise-concepts/two-fer.md [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+the exported function. Most of the time they leave the stub comments in the file.
+
+### Functions
+
+Everything is done via functions. In this case, it's also:
+
+- there is no concept of default values for paramaters, it's done in the body of the function
```
Comment (bitfield) [inline languages/go/reference/exercise-concepts/two-fer.md]:
```suggestion
- there is no concept of default values for parameters; it's done in the body of the function
```

File: languages/go/reference/exercise-concepts/two-fer.md [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+- parameters are passed by-value, so in the scope of the function they can be altered
+
+### Strings
+
+The parameter passed is a string. For the default value, a check needs to be made if the string is empty.
+
+- how to do a idiomatic check on empty string
```
Comment (bitfield) [inline languages/go/reference/exercise-concepts/two-fer.md]:
```suggestion
- how to do an idiomatic check for empty string
```

File: languages/go/reference/exercise-concepts/two-fer.md [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+- how to concatenate strings, either via simple `+` or via the fmt package
```
Comment (bitfield) [inline languages/go/reference/exercise-concepts/two-fer.md]:
```suggestion
- how to concatenate strings, either using a simple `+`, or the `fmt` package
```

File: languages/go/reference/exercise-concepts/two-fer.md (hunk 6 above)
Comment (bitfield) [inline languages/go/reference/exercise-concepts/two-fer.md]:
```suggestion
The code needs to check whether the parameter is empty.
```

**Comment by tehsphinx** [comment]
I rebased this so the link-checkers work now

File: languages/go/reference/exercise-concepts/two-fer.md (hunk 1 above)
Comment (tehsphinx) [inline languages/go/reference/exercise-concepts/two-fer.md]:
```suggestion
- parameters are passed by value, so in the scope of the function they can be altered
```

**Comment by tehsphinx** [comment]
Just as a side node: These `concepts needed for v2`-issues (and PRs) are there to figure out which concepts exists to complete out list of concepts: https://github.com/exercism/v3/issues/167.

And they will help us later to define the prerequisites of a practice exercise (v2 exercise) which in turn defines the tree of exercises that needs to be finished before a practice exercise will be unlocked.

What these PRs are not for is to determine how to solve an exercise in an idiomatic way. It's not about right vs wrong, good vs bad, fast vs slow, etc. It's purely about which concepts are (commonly) used to solve this

**Pull request merged**